File: tools-manage/invocationlab-admin/src/main/java/org/javamaster/invocationlab/admin/service/AppFactory.java
package org.javamaster.invocationlab.admin.service;

import org.javamaster.invocationlab.admin.enums.RegisterCenterType;
import org.javamaster.invocationlab.admin.service.creation.Creator;
import org.javamaster.invocationlab.admin.service.creation.impl.DubboCreator;
import org.javamaster.invocationlab.admin.service.creation.impl.EurekaCreator;
import org.javamaster.invocationlab.admin.service.invocation.Invoker;
import org.javamaster.invocationlab.admin.service.invocation.entity.PostmanDubboRequest;
import org.javamaster.invocationlab.admin.service.registry.RegisterFactory;
import org.javamaster.invocationlab.admin.service.registry.impl.DubboRegisterFactory;
import org.javamaster.invocationlab.admin.service.registry.impl.EurekaRegisterFactory;
import org.javamaster.invocationlab.admin.service.repository.redis.RedisKeys;
import org.javamaster.invocationlab.admin.service.repository.redis.RedisRepository;
import org.javamaster.invocationlab.admin.util.SpringUtils;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.beans.factory.annotation.Qualifier;
import org.springframework.stereotype.Component;

/**
 * @author yudong
 * @date 2022/11/12
 */
@Component
public class AppFactory {
    @Autowired
    private RedisRepository redisRepository;
    @Autowired
    private Invoker<Object, PostmanDubboRequest> dubboInvoker;
    @Autowired
    @Qualifier("feignInvoker")
    private Invoker<Object, PostmanDubboRequest> feignInvoker;

    public static RegisterFactory getRegisterFactory(Integer type) {
        RegisterCenterType registrationCenterType = RegisterCenterType.getByType(type);
        if (registrationCenterType == RegisterCenterType.ZK) {
            return SpringUtils.getContext().getBean(DubboRegisterFactory.class);
        } else if (registrationCenterType == RegisterCenterType.EUREKA) {
            return SpringUtils.getContext().getBean(EurekaRegisterFactory.class);
        }
        throw new RuntimeException("wrong type:" + type);
    }

    public RegisterFactory getRegisterFactory(String zk) {
        Integer type = redisRepository.mapGet(RedisKeys.CLUSTER_REDIS_KEY_TYPE, zk);
        return getRegisterFactory(type);
    }

    public Invoker<Object, PostmanDubboRequest> getInvoker(String cluster) {
        Integer type = redisRepository.mapGet(RedisKeys.CLUSTER_REDIS_KEY_TYPE, cluster);
        RegisterCenterType centerType = RegisterCenterType.getByType(type);
        if (centerType == RegisterCenterType.ZK) {
            return dubboInvoker;
        } else if (centerType == RegisterCenterType.EUREKA) {
            return feignInvoker;
        } else {
            throw new RuntimeException("wrong type:" + type);
        }
    }

    public RegisterCenterType getType(String cluster) {
        Integer type = redisRepository.mapGet(RedisKeys.CLUSTER_REDIS_KEY_TYPE, cluster);
        return RegisterCenterType.getByType(type);
    }

    public Creator getCreator(String zk) {
        Integer type = redisRepository.mapGet(RedisKeys.CLUSTER_REDIS_KEY_TYPE, zk);
        RegisterCenterType registrationCenterType = RegisterCenterType.getByType(type);
        if (registrationCenterType == RegisterCenterType.ZK) {
            return SpringUtils.getContext().getBean(DubboCreator.class);
        } else if (registrationCenterType == RegisterCenterType.EUREKA) {
            return SpringUtils.getContext().getBean(EurekaCreator.class);
        }
        throw new RuntimeException("wrong type:" + type);
    }
}

File: tools-manage/invocationlab-admin-21/src/main/kotlin/org/javamaster/invocationlab/admin/service/AppFactory.kt
package org.javamaster.invocationlab.admin.service

import org.javamaster.invocationlab.admin.enums.RegisterCenterType
import org.javamaster.invocationlab.admin.enums.RegisterCenterType.Companion.getByType
import org.javamaster.invocationlab.admin.service.creation.Creator
import org.javamaster.invocationlab.admin.service.creation.impl.DubboCreator
import org.javamaster.invocationlab.admin.service.creation.impl.EurekaCreator
import org.javamaster.invocationlab.admin.service.invocation.Invoker
import org.javamaster.invocationlab.admin.service.invocation.entity.PostmanDubboRequest
import org.javamaster.invocationlab.admin.service.registry.RegisterFactory
import org.javamaster.invocationlab.admin.service.registry.impl.DubboRegisterFactory
import org.javamaster.invocationlab.admin.service.registry.impl.EurekaRegisterFactory
import org.javamaster.invocationlab.admin.service.repository.redis.RedisKeys
import org.javamaster.invocationlab.admin.service.repository.redis.RedisRepository
import org.javamaster.invocationlab.admin.util.SpringUtils
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.beans.factory.annotation.Qualifier
import org.springframework.stereotype.Component

/**
 * @author yudong
 * @date 2022/11/12
 */
@Component
class AppFactory {
    @Autowired
    private lateinit var redisRepository: RedisRepository

    @Autowired
    private lateinit var dubboInvoker: Invoker<Any, PostmanDubboRequest>

    @Autowired
    @Qualifier("feignInvoker")
    private lateinit var feignInvoker: Invoker<Any, PostmanDubboRequest>

    fun getRegisterFactory(zk: String): RegisterFactory {
        val type = redisRepository.mapGet<Int>(RedisKeys.CLUSTER_REDIS_KEY_TYPE, zk)
        return getRegisterFactory(type!!)
    }

    fun getInvoker(cluster: String): Invoker<Any, PostmanDubboRequest> {
        val type = redisRepository.mapGet<Int>(RedisKeys.CLUSTER_REDIS_KEY_TYPE, cluster)
        val centerType = getByType(type)
        return when (centerType) {
            RegisterCenterType.ZK -> {
                dubboInvoker
            }

            RegisterCenterType.EUREKA -> {
                feignInvoker
            }

            else -> {
                throw RuntimeException("wrong type:$type")
            }
        }
    }

    fun getType(cluster: String): RegisterCenterType? {
        val type = redisRepository.mapGet<Int>(RedisKeys.CLUSTER_REDIS_KEY_TYPE, cluster)
        return getByType(type)
    }

    fun getCreator(zk: String): Creator {
        val type = redisRepository.mapGet<Int>(RedisKeys.CLUSTER_REDIS_KEY_TYPE, zk)
        val registrationCenterType = getByType(type)
        if (registrationCenterType == RegisterCenterType.ZK) {
            return SpringUtils.context.getBean(DubboCreator::class.java)
        } else if (registrationCenterType == RegisterCenterType.EUREKA) {
            return SpringUtils.context.getBean(EurekaCreator::class.java)
        }
        throw RuntimeException("wrong type:$type")
    }

    companion object {
        @JvmStatic
        fun getRegisterFactory(type: Int): RegisterFactory {
            val registrationCenterType = getByType(type)
            if (registrationCenterType == RegisterCenterType.ZK) {
                return SpringUtils.context.getBean(DubboRegisterFactory::class.java)
            } else if (registrationCenterType == RegisterCenterType.EUREKA) {
                return SpringUtils.context.getBean(EurekaRegisterFactory::class.java)
            }
            throw RuntimeException("wrong type:$type")
        }
    }
}
